core/src/typecheck/mod.rs

```diff
@@ -2204,18 +2237,35 @@ fn infer_with_annot<L: Linearizer>(
 /// `infer` corresponds to the inference mode of bidirectional typechecking. Nickel uses a mix of
 /// bidirectional typechecking and traditional ML-like unification.
 fn infer<L: Linearizer>(
+    state: &mut State,
+    ctxt: Context,
+    lin: &mut Linearization<L::Building>,
+    linearizer: L,
+    rt: &RichTerm,
+) -> Result<UnifType, TypecheckError> {
+    infer_visited(state, ctxt, lin, linearizer, rt, None)
+}
+
+/// Variant of [infer] taking an additional `item_id` to indicate that inference is done on a term
+/// that has already been visited by the linearizer, for example when coming directly from [check].
+/// When `item_id` is defined, the linearizer isn't called. [infer] is just `infer_visited` called
+/// with `item_id` set to `None`.
+///
+/// Whether `item_id` is `Some` or `None`, the item is always amended with its inferred type.
+fn infer_visited<L: Linearizer>(
     state: &mut State,
     mut ctxt: Context,
     lin: &mut Linearization<L::Building>,
     mut linearizer: L,
     rt: &RichTerm,
+    item_id: Option<L::ItemId>,
 ) -> Result<UnifType, TypecheckError> {
     let RichTerm { term, pos } = rt;
 
-    // For the following infer rules, we need to be careful to call to `linearizer.add_term` at the
-    // right moment. We need to provide a type, so we can't do it before having inferred a type for
-    // `rt`. However, we must do it **before** any subsequent calls to `check` or `infer`, because
-    // the linearizer relies on terms being visited in order (pre-order).
+    // We don't have a type yet, so we put `Dyn` for now, but we'll update once we infer one via
+    // `linearizer.retype`
+    let item_id = item_id.or_else(|| linearizer.add_term(lin, rt, mk_uniftype::dynamic()));
```

Comment:
I'm a little worried about the fact that `item_id == None` could have two meanings: it wasn't added to the linearizer, or it was added but the linearizer returned `None`. In the second case, we could still call `add_term` twice, right?

Reply:
Technically, yes. Adding such a term twice is currently not an issue, because if the linearizer returned `None` once, it'll always return `None` (the term doesn't have a position). On those terms `add_term` is basically a no-op. Though it's a bit fragile and specific to the current linearizer implementation.

I'll try to think if there's a simple change that makes it better, or if that'll do for now as we're going to scrap the linearizer soon anyway.

Reply:
If it doesn't break anything right now, I'd be inclined to just not worry :shrug: